### PR TITLE
Fix compiler errors on Windows

### PR DESCRIPTION
**Pull request by yocabon**
## What does this PR do?
Fixes # (issue).

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - https://github.com/facebookresearch/xformers/issues/437
- [x] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [x] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: setup.py
Comment (fmassa) [inline setup.py:116]: Can this be conditionned on Windows builds only?

File: xformers/components/attention/csrc/attention.cpp
Comment (fmassa) [inline xformers/components/attention/csrc/attention.cpp:14]: Thanks for this! This is a longstanding issue with torchvision as well which we haven't found a better solution so far.

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h
Comment (fmassa) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h:1358]: @danthe3rd looks like we might have switched the order of the `restrict` keywords here. Do you think the `__restrict__` keyword was being picked by the compiler before in here?

**Review by fmassa — APPROVED** [review]
This all LGTM, thanks a ton @yocabon !

I only have really one request -- could you try making C++17 be a dependency only on Windows for flashattention?

File: setup.py
Comment (yocabon) [reply setup.py]: https://github.com/facebookresearch/xformers/pull/460/commits/bf5b1fd922bec9a7723723e1d014e7d09a60fcea should do it

File: xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu
Comment (danthe3rd) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu:158]: as this is just a sanity check - can we keep it as it was but under an `#ifdef` guard for windows?

Comment (yocabon) [reply xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu]: https://github.com/facebookresearch/xformers/pull/460/commits/8509baaef125121c4cf7f5476eb4c26e1d0769c6 should do it

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h
Comment (danthe3rd) [reply xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h]: that's a good point. I'll run some benchmarks after that

**Review by danthe3rd — APPROVED** [review]
Thanks a lot! That's amazing!
Let's wait for the CI before merging. I believe there might be a merge conflict also ...

**Pull request merged**
